[HigherOrderOp] Flatten outputs of `wrap`.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110417
* #107618
* __->__ #109433
* #110290

Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ysiraichi committed:
[HigherOrderOp] Flatten outputs of `wrap`.

Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Fix + rebase. on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1097,6 +1104,7 @@ def create_wrapped_node(self, tx, args, kwargs, description):
             checkpoint,
             description,
             manually_set_subgraph_inputs=False,
+            should_flatten_outputs=True,
```
Here's what we should do in WrapHigherOrderVariable.call_function:
- it calls speculate_subgraph with should_flatten_outputs=True
- it writes a call to `outputs = wrap(body_function, *args)`
- finally, it unflattens the output variables and returns that

In order to unwrap the output variables: perhaps we could construct a TorchVariable that is the pytree.tree_unflatten and invoke .call_function ?

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Fix linting issues. on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented:
@zou3519 This is a friendly reminder. Do you have some time to take a look at it?

zou3519 commented on test/dynamo/test_subclasses.py:
```diff
@@ -312,19 +312,20 @@ def forward(self, L_x_ : torch.Tensor):
         ff_out = ff(t_clone)
         # frame count and op count are incremented due to re-compilation
-        check_count_and_graph(2, 2, 2, expected_graph)
+        check_count_and_graph(2, 4, 2, expected_graph)
```
What is the middle number? Why did it go up?

ysiraichi replied:
It's the number of expected operations. What's happening there is:
- `wrap` returns a tuple (consistent with plain dynamo)
- A new node `call_function(get_item)` is needed to create a proxy for the inner `TensorVariable` (+1 extra op)
So, we are doubling the number of expected operations if `wrap` returns a single tensor.

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -120,7 +120,7 @@ def _assert_wrap_fallback(self, func, args, setup=lambda: None):
         self.assertEqual(result, expected)
 
-    def _test_wrap_simple(self, func, args, expected_num_wrap_args, expected_opcount=1):
+    def _test_wrap_simple(self, func, args, expected_num_wrap_args, expected_opcount=2):
```
Could we add a self.assertExpectedInline test on the graph to confirm that our outputs have been flattened? One way to do this is to change `_test_wrap_simple` to return backend.graphs and then self.assertExpectedInline on the return in one of our tests

zou3519 replied:
Maybe wait for https://github.com/pytorch/pytorch/pull/109962 which does this already

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -867,14 +870,14 @@ def fn(*, x, y, z=None):
         opt = torch.compile(f, backend="eager", fullgraph=True)
         opt(x, y)
-        self.assertEqual(counters["stats"]["calls_captured"], 1)
+        self.assertEqual(counters["stats"]["calls_captured"], 2)
```
To check my understanding: this changed from 1->2 because we now have a call turning the output into a Tuple?

ysiraichi replied:
See [this comment](https://github.com/pytorch/pytorch/pull/109433#discussion_r1337506405).

zou3519 reviewed (commented):
approach LGTM. We should add at least one test with assertExpectedInline so we can visualize the output. Some minor questions about what is going on in the tests.

ysiraichi committed:
Rebased + CI fix. on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Using PyTree tracing. on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented:
@zou3519 I think this PR is ready for review. I have rebased it to #109962 commit, so we have `assertExpectedInline` tests. Could you take a look at it?

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1631,31 +1637,26 @@ def f(x):
         self.assertEqual(result, x.sin() + x.cos())
         self.assertEqual(cnt.frame_count, 1)
+        self.assertEqual(len(counters["graph_break"]), 0)
         self.assertEqual(len(backend.graphs), 1)
```
can we do a self.assertExpectedInline on backend.graphs[0]? We should see that the wrap's outputs are flattened

zou3519 commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1631,31 +1637,26 @@ def f(x):
-        )
+        self.assertEqual(result[0]["a"], -x)
+        self.assertEqual(cnt.frame_count, 1)
+        self.assertEqual(len(counters["graph_break"]), 0)
```
Can we acquire the graph here and do a self.assertExpectedInlne here? We should see that the output of wrap was flattened

zou3519 reviewed (commented):
code looks fine, i want to see an assertExpectedInline on a graph for a function where the f passed to wrap returns nested tuples or a dict

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 reviewed (approved):
perfect, thank you

ysiraichi commented:
CI failure seems to be coming from upstream.

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented:
CI failures seem to be coming from upstream.

ysiraichi commented:
@pytorchbot merge

kit1980 commented:
> CI failures seem to be coming from upstream.

@ysiraichi CI failures are from the previous PR in the stack.

kit1980 commented:
@pytorchbot revert -m "Depends on reverted https://github.com/pytorch/pytorch/pull/110290" -c ignoredsignal

kit1980 commented:
> @pytorchbot revert -m "Depends on reverted #110290" -c ignoredsignal

Actually the issue was introduced by this PR, not the previous one in the stack #110290

ysiraichi committed:
Rebased. on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ysiraichi commented:
I think it was a landing conflict. A rebase should fix that.

ysiraichi committed:
Update on "[HigherOrderOp] Flatten outputs of `wrap`."


Fix: #109247

This PR flattens `wrap` outputs by inlining `pytree.tree_flatten` function after calling
the inner function.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "unrelated failures"
